dynamic_programming/word_break.py

```diff
@@ -0,0 +1,25 @@
+def wordbreak(word: str, word_dict: list) -> bool:
+    """
+    word ---> Input word
+    word_dict ---> Input dictionary
+    >>> wordbreak('deepImpact',['deep','Im','pact'])
+    True
+    >>> wordbreak('mangoinmango',['mango','in'])
+    True
+    >>> wordbreak('deepfacts',['deeps','fact'])
+    False
+    """
+    dict_set = set(word_dict)
+    start = [0]
+    # start[0] is the equivalent of a base-case
+    # from the recursive solution and start[-1] is the
+    # overall solution to the complete problem
+    for i in range(len(word)):
+        for j in start:
+            if word[j : i + 1] in dict_set:
+                start.append(i + 1)  # start of next word in dict

```

Comment:
I don't like the idea of appending to the list that is being currently iterated through. Should we copy the start list when iterating over it?

Reply:
I don't think appending to `start` while iterating over it should pose any issues because we immediately break out of that for-loop when we append. This bottom-up DP approach normally involves instantiating the whole array at the beginning, so appending to the list like this usually never happens. I'm guessing that @samyak1999 chose to implement it using a dynamic array in order to minimize the number of indices that need to be iterated over.